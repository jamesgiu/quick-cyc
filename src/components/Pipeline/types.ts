export enum PipelineIntent {
  NONE = "qc-intent-none",
  WARNING = "qc-intent-warning",
  SUCCESS = "qc-intent-success",
  FAILURE = "qc-intent-failure",
  IN_PROGRESS = "qc-intent-in-progress",
  SKIPPED = "qc-intent-skipped",
}

export enum PipelineSize {
  XS = "qc-size-xs",
  S = "qc-size-s",
  M = "qc-size-m",
  L = "qc-size-l",
  XL = "qc-size-xl",
}

export enum PipelineIntentColours {
  NONE = "#5f6b7c",
  WARNING = "#fbb360",
  SUCCESS = "#238551",
  FAILURE = "#cd4246",
  IN_PROGRESS = "#2d72d2",
  SKIPPED = "#75787c",
}
