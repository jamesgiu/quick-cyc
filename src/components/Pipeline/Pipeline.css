.qc-pipeline {
  display: flex;
  flex-direction: row;
}

.qc-pipeline .qc-pipe-wrapper {
  margin: auto;
  margin-left: -10px;
  margin-right: -10px;
  z-index: 0;
}

.qc-subpipeline .qc-node-wrapper {
  margin: auto;
}

.qc-subpipeline {
  border: dashed 0.1px;
  margin-top: auto;
  margin-bottom: auto;
}
