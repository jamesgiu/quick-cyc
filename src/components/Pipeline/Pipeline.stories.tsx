import type { Meta, StoryObj } from '@storybook/react';

import { Pipeline } from './Pipeline';
import { PipelineIcons } from './icons';
import { PipelineIntent, PipelineSize } from './types';

const meta: Meta<typeof Pipeline> = {
    title: 'Pipeline',
    component: Pipeline,
    parameters: {
      // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
      layout: 'centered',
    },
    // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
    tags: ['autodocs']
  };

type Story = StoryObj<typeof Pipeline>;
  
export const Primary: Story = {
    args: {
      label: 'Pipeline',
      schema: [
        {
        icon:PipelineIcons.Icon2fa,
        intent:PipelineIntent.SUCCESS,
        active:false,
        attempts:1,
        size:PipelineSize.M,
        },
        {
          intent:PipelineIntent.SUCCESS,
          active:false,
          size:PipelineSize.S,
        },
        {
          icon:PipelineIcons.IconAward,
          intent:PipelineIntent.SUCCESS,
          active:false,
          attempts:1,
          size:PipelineSize.M,
        },
        {
          intent:PipelineIntent.SUCCESS,
          active:false,
          size:PipelineSize.S,
        },
        {
          icon:PipelineIcons.IconClock,
          intent:PipelineIntent.IN_PROGRESS,
          active:true,
          size:PipelineSize.M,
        },
        {
          intent:PipelineIntent.IN_PROGRESS,
          active:true,
          progressPercent: 30,
          size:PipelineSize.S,
        },
        {
          icon:PipelineIcons.Icon3dRotate,
          intent:PipelineIntent.NONE,
          active:false,
          size:PipelineSize.M,
        },
      ]
    }
};

export const SkipPipeline: Story = {
  args: {
    label: 'Pipeline with skips',
    schema: [
      {
      icon:PipelineIcons.Icon2fa,
      intent:PipelineIntent.SUCCESS,
      active:false,
      attempts:1,
      size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.SKIPPED,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconAward,
        intent:PipelineIntent.SKIPPED,
        active:false,
        attempts:0,
        size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.SKIPPED,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconClock,
        intent:PipelineIntent.IN_PROGRESS,
        active:true,
        size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.IN_PROGRESS,
        active:true,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.Icon3dRotate,
        intent:PipelineIntent.NONE,
        active:false,
        size:PipelineSize.M,
      },
    ]
  }
};

export const Subpipeline: Story = {
  args: {
    label: 'Pipeline',
    schema: [
      {
      icon:PipelineIcons.Icon2fa,
      intent:PipelineIntent.SUCCESS,
      active:false,
      attempts:1,
      size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.SUCCESS,
        active:false,
        size:PipelineSize.S,
      },
      {
        label: "Subpipeline",
        schema: [
          {
            icon:PipelineIcons.IconBrandAws,
            intent:PipelineIntent.SUCCESS,
            active:false,
            attempts:1,
            size:PipelineSize.S,
            },
            {
              intent:PipelineIntent.SUCCESS,
              active:false,
              size:PipelineSize.XS,
            },
            {
              icon:PipelineIcons.IconClick,
              intent:PipelineIntent.SUCCESS,
              active:false,
              attempts:1,
              size:PipelineSize.S,
            }
        ],
      },
      {
        intent:PipelineIntent.SUCCESS,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconAward,
        intent:PipelineIntent.WARNING,
        active:true,
        attempts:1,
        size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.WARNING,
        active:true,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconClock,
        intent:PipelineIntent.NONE,
        active:false,
        attempts:30,
        size:PipelineSize.M,
      },
      {
        intent:PipelineIntent.NONE,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.Icon3dRotate,
        intent:PipelineIntent.NONE,
        active:false,
        size:PipelineSize.M,
      },
    ]
  }
};

export const Annotated: Story = {
  args: {
    label: 'Annotated pipeline',
    schema: [
      {
      icon:PipelineIcons.Icon2fa,
      intent:PipelineIntent.SUCCESS,
      active:false,
      attempts:1,
      size:PipelineSize.M,
      outerLabel: "Auth",
      },
      {
        intent:PipelineIntent.SUCCESS,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconAward,
        intent:PipelineIntent.SUCCESS,
        active:false,
        attempts:1,
        size:PipelineSize.M,
        outerLabel: "Verify",
      },
      {
        intent:PipelineIntent.SUCCESS,
        active:false,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.IconClock,
        intent:PipelineIntent.IN_PROGRESS,
        active:true,
        size:PipelineSize.M,
        outerLabel: "Sleep",
      },
      {
        intent:PipelineIntent.IN_PROGRESS,
        active:true,
        progressPercent: 30,
        size:PipelineSize.S,
      },
      {
        icon:PipelineIcons.Icon3dRotate,
        intent:PipelineIntent.NONE,
        active:false,
        size:PipelineSize.M,
        outerLabel: "Restore",
      },
    ]
  }
};

  
      
export default meta;