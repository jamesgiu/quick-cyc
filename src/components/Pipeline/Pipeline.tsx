import React from "react";
import "./Pipeline.css";
import "../index.css";
import Pipe, {
  PipelinePipeProps as PipelinePipeProps,
} from "./shared/PipelinePipe/PipelinePipe";
import {
  PipelineNode,
  PipelineNodeProps,
} from "./shared/PipelineNode/PipelineNode";

export interface PipelineProps {
  label: string;
  schema: Array<PipelineNodeProps | PipelinePipeProps | PipelineProps>;
}

const renderPipeline = (
  schema: Array<PipelineNodeProps | PipelinePipeProps | PipelineProps>,
): JSX.Element[] => {
  const elements: JSX.Element[] = [];

  schema.forEach((item) => {
    // Ensure we are dealing with a node, otherwise, it's a pipe (// FIXME bit hacky)
    if (Object.keys(item).includes("icon")) {
      const node = item as PipelineNodeProps;
      elements.push(
        <PipelineNode
          className={node.className}
          icon={node.icon}
          intent={node.intent}
          onNodeClick={node.onNodeClick}
          attempts={node.attempts}
          percentComplete={node.percentComplete}
          size={node.size}
          active={node.active}
          innerLabel={node.innerLabel}
          outerLabel={node.outerLabel}
        />,
      );
    } else if (Object.keys(item).includes("schema")) {
      const pipeline = item as PipelineProps;
      elements.push(
        <span className={"qc-subpipeline"}>
          <Pipeline label="embedded" schema={pipeline.schema} />
        </span>,
      );
    } else {
      const pipe = item as PipelinePipeProps;
      elements.push(
        <Pipe
          intent={pipe.intent}
          size={pipe.size}
          active={pipe.active}
          progressPercent={pipe.progressPercent}
        />,
      );
    }
  });

  return elements;
};

export const Pipeline = (props: PipelineProps) => {
  return <span className={"qc-pipeline"}>{renderPipeline(props.schema)}</span>;
};

export default Pipeline;
