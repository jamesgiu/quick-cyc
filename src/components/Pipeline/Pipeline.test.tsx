// @vitest-environment jsdom
 
import { expect, test } from 'vitest';
 
// Replace your-framework with the framework you are using, e.g. react-vite, nextjs, nextjs-vite, etc.
import { composeStories } from '@storybook/react-vite';
 
import * as stories from './Pipeline.stories';
 
const { Primary } = composeStories(stories);
test('Pipeline snapshot', async () => {
  await Primary.run();
  expect(document.body.firstChild).toMatchSnapshot();
});