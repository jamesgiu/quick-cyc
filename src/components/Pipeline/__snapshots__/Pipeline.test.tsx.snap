// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pipeline snapshot 1`] = `
<div>
  <span
    class="qc-pipeline"
  >
    <span
      class="qc-node-wrapper undefined qc-size-m qc-intent-success  "
    >
      <span
        class="qc-node "
      >
        <span
          class="qc-top-right-icon qc-size-m qc-intent-success"
        >
          1
        </span>
        <span
          class="qc-node-icon"
        >
          <svg
            class="tabler-icon tabler-icon-2fa"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M7 16h-4l3.47 -4.66a2 2 0 1 0 -3.47 -1.54"
            />
            <path
              d="M10 16v-8h4"
            />
            <path
              d="M10 12l3 0"
            />
            <path
              d="M17 16v-6a2 2 0 0 1 4 0v6"
            />
            <path
              d="M17 13l4 0"
            />
          </svg>
        </span>
      </span>
    </span>
    <div
      class="qc-pipe-wrapper qc-size-s qc-intent-success "
    >
      <div
        class="qc-pipe-loading-wrapper"
      >
        <div
          class="qc-pipe"
        />
      </div>
    </div>
    <span
      class="qc-node-wrapper undefined qc-size-m qc-intent-success  "
    >
      <span
        class="qc-node "
      >
        <span
          class="qc-top-right-icon qc-size-m qc-intent-success"
        >
          1
        </span>
        <span
          class="qc-node-icon"
        >
          <svg
            class="tabler-icon tabler-icon-award"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 9m-6 0a6 6 0 1 0 12 0a6 6 0 1 0 -12 0"
            />
            <path
              d="M12 15l3.4 5.89l1.598 -3.233l3.598 .232l-3.4 -5.889"
            />
            <path
              d="M6.802 12l-3.4 5.89l3.598 -.233l1.598 3.232l3.4 -5.889"
            />
          </svg>
        </span>
      </span>
    </span>
    <div
      class="qc-pipe-wrapper qc-size-s qc-intent-success "
    >
      <div
        class="qc-pipe-loading-wrapper"
      >
        <div
          class="qc-pipe"
        />
      </div>
    </div>
    <span
      class="qc-node-wrapper undefined qc-size-m qc-intent-in-progress qc-state-active "
    >
      <span
        class="qc-node "
      >
        <span
          class="qc-node-icon"
        >
          <svg
            class="tabler-icon tabler-icon-clock"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M3 12a9 9 0 1 0 18 0a9 9 0 0 0 -18 0"
            />
            <path
              d="M12 7v5l3 3"
            />
          </svg>
        </span>
      </span>
    </span>
    <div
      class="qc-pipe-wrapper qc-size-s qc-intent-in-progress qc-state-active"
    >
      <div
        class="qc-pipe-loading-wrapper"
        style="width: 30%;"
      >
        <div
          class="qc-pipe"
        />
      </div>
    </div>
    <span
      class="qc-node-wrapper undefined qc-size-m qc-intent-none  "
    >
      <span
        class="qc-node "
      >
        <span
          class="qc-node-icon"
        >
          <svg
            class="tabler-icon tabler-icon-3d-rotate"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 3a7 7 0 0 1 7 7v4l-3 -3"
            />
            <path
              d="M22 11l-3 3"
            />
            <path
              d="M8 15.5l-5 -3l5 -3l5 3v5.5l-5 3z"
            />
            <path
              d="M3 12.5v5.5l5 3"
            />
            <path
              d="M8 15.545l5 -3.03"
            />
          </svg>
        </span>
      </span>
    </span>
  </span>
</div>
`;
